linera-views/src/shared_view.rs

```diff
@@ -0,0 +1,220 @@
+                have already finished"
+            );
+        }
+
+        *writer_notifier = Some(sender);
+
+        Ok(ReadWriteViewReference {
+            view: self.view.share_unchecked()?,
+            write_barrier: Some(receiver),
+            _writer_permit: writer_permit,
+        })
+    }
+}
+
+/// A read-only reference to a [`SharedView`].
+pub struct ReadOnlyViewReference<V> {
+    view: V,
+    reader_count: Arc<AtomicUsize>,
+    writer_notifier: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+}
+
+impl<V> Deref for ReadOnlyViewReference<V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        &self.view
+    }
+}
+
+impl<V> Drop for ReadOnlyViewReference<V> {
+    fn drop(&mut self) {
+        if self.reader_count.fetch_sub(1, Ordering::SeqCst) == 1 {
+            let mut writer_notifier = self
+                .writer_notifier
+                .lock()
+                .expect("No panics should happen while holding the `write_notifier` lock");
+
+            if self.reader_count.load(Ordering::Acquire) == 0 {
+                if let Some(notifier) = writer_notifier.take() {
+                    let _ = notifier.send(());
+                }
+            }
+        }
+    }
+}
+
+/// A read-write reference to a [`SharedView`].
+pub struct ReadWriteViewReference<V> {
+    view: V,
+    write_barrier: Option<oneshot::Receiver<()>>,
+    _writer_permit: SemaphoreGuardArc,
+}
+
+impl<V> Deref for ReadWriteViewReference<V> {
+    type Target = V;
+
+    fn deref(&self) -> &Self::Target {
+        &self.view
+    }
+}
+
+impl<V> DerefMut for ReadWriteViewReference<V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.view
+    }
+}
+
+#[async_trait]
+impl<C, V> View<C> for ReadWriteViewReference<V>
+where
+    C: Send + 'static,
+    V: View<C>,
+{
+    fn context(&self) -> &C {
+        self.deref().context()
+    }
+
+    async fn load(_context: C) -> Result<Self, ViewError> {
+        unreachable!("`ReadWriteViewReference` should not be loaded directly");
+    }
+
+    fn rollback(&mut self) {
+        self.deref_mut().rollback();
+    }
+
+    fn clear(&mut self) {
+        self.deref_mut().clear();
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<(), ViewError> {
+        self.deref_mut().flush(batch)
+    }
+
+    fn share_unchecked(&mut self) -> Result<Self, ViewError> {
+        unreachable!(
+            "`ReadWriteViewReference` should not be shared without going through its parent \
+            `SharedView`"
+        );
+    }
+}
+
+#[async_trait]
+impl<C, V> RootView<C> for ReadWriteViewReference<V>
+where
+    C: Context + Send + 'static,
+    V: View<C> + Send,
+    ViewError: From<C::Error>,
+{
+    async fn save(&mut self) -> Result<(), ViewError> {
+        if let Some(barrier) = self.write_barrier.take() {
+            let () = barrier
+                .await
+                .expect("`writer_notifier` should never be dropped without notifying first");
+        }
```

Comment:
Ok, a write barrier is created. But the writing itself occurs outside of that barrier. So, it does not seem right.